File: src/main/java/network/Network.java
package network;

import guru.nidi.graphviz.attribute.Color;
import guru.nidi.graphviz.attribute.Label;
import guru.nidi.graphviz.attribute.Style;
import guru.nidi.graphviz.model.MutableGraph;
import guru.nidi.graphviz.model.MutableNode;
import graph.algorithm.DejkstraFactory;
import graph.algorithm.GraphPath;
import graph.algorithm.Algorithm;
import graph.algorithm.Factory;
import graph.edge.Edge;
import graph.vertex.Vertex;
import utils.Stringable;
import utils.PropertiesHandler;
import org.jetbrains.annotations.NotNull;

import java.util.*;
import java.util.stream.Collectors;

import static guru.nidi.graphviz.model.Factory.*;

public class Network implements Stringable {
    private List<Vertex> nodes;
    private final Topology topology;
    private Factory factory;
    private List<GraphPath> openConnections;

    public Network() {
        this(new Topology() {
            @Override
            public List<Vertex> createNetwork() {
                return new ArrayList<>();
            }

            @Override
            public Vertex createNode() {
                return null;
            }
        });
    }

    public Network(@NotNull Topology topology) {
        this(topology, new DejkstraFactory());
    }

    public Network(@NotNull Topology topology, Factory factory) {
        this.topology = topology;
        this.factory = factory;
        generateNetwork();
    }

    public void generateNetwork() {
        nodes = topology.createNetwork();
        openConnections = new ArrayList<>();
    }

    public List<Vertex> getNodes() {
        return nodes;
    }

    public List<Vertex> getNodes(boolean terminal) {
        if (!terminal)
            return nodes;
        return nodes.stream()
                .filter(Vertex::isTerminal)
                .collect(Collectors.toList());
    }

    public Vertex getRandomNode() {
        return nodes.get(new Random().nextInt(nodes.size()));
    }

    public Vertex getRandomNode(boolean terminal) {
        List<Vertex> nodes = getNodes(terminal);
        return nodes.get(new Random().nextInt(nodes.size()));
    }

    public Factory getFactory() {
        return factory;
    }

    public void setFactory(Factory factory) {
        this.factory = factory;
    }

    public GraphPath getPath(@NotNull Vertex from, @NotNull Vertex to) {
        return getPath(from, to, false);
    }

    public GraphPath getPath(@NotNull Vertex from, @NotNull Vertex to, boolean isUsed) {
        if (!exists(from, to))
            return new GraphPath();
        return factory.getAlgorithm(nodes, getChannels(isUsed)).getPath(from, to);
    }

    public List<GraphPath> getPaths(@NotNull Vertex from) {
        Algorithm algorithm = factory.getAlgorithm(nodes, getChannels());
        return nodes.stream()
                .filter(node -> !node.equals(from))
                .map(node -> algorithm.getPath(from, node))
                .collect(Collectors.toList());
    }

    private List<Edge> getChannels() {
        return nodes.stream()
                .flatMap(node -> node.getConnections().stream())
                .distinct()
                .collect(Collectors.toList());
    }

    private List<Edge> getChannels(boolean isUsed) {
        return nodes.stream()
                .flatMap(node -> node.getConnections().stream())
                .distinct()
                .filter(ch -> ch.isUsed() == isUsed)
                .collect(Collectors.toList());
    }

    public GraphPath createConnection(@NotNull Vertex from, @NotNull Vertex to) {
        if (from == to)
            return new GraphPath();
        GraphPath path = getPath(from, to);
        if (path.exists()) {
            path.forEach(ch -> ch.setUsed(true));
            openConnections.add(path);
        }
        return path;
    }

    public void closeConnection(@NotNull GraphPath path) {
        if (openConnections.contains(path)) {
            path.forEach(ch -> ch.setUsed(false));
            openConnections.remove(path);
        }
    }


    public void closeConnection(@NotNull Vertex from, @NotNull Vertex to) {
        openConnections.stream()
                .filter(path -> path.ofNodes(from, to))
                .findAny()
                .ifPresent(this::closeConnection);
    }

    public Vertex addNode() {
        Vertex node = topology.createNode();
        if (node != null) {
            nodes.add(node);
        }
        return node;
    }

    public void addNode(@NotNull Vertex node) {
        if (!exists(node)) {
            nodes.add(node);
        }
    }

    public void removeNode(@NotNull Vertex node) {
        Edge[] connections = node.getConnections().toArray(new Edge[]{});
        for (Edge ch: connections) {
            ch.remove();
        }
        nodes.remove(node);
    }

    public void addConnection(@NotNull Vertex from, @NotNull Vertex to, int weight) {
        if (exists(from, to)) {
            from.addConnectedNode(to, weight);
        }
    }

    public void removeConnection(@NotNull Vertex from, @NotNull Vertex to) {
        from.removeConnectedNode(to);

    }

    private boolean exists(Vertex... nodes) {
        return Arrays.stream(nodes)
                .allMatch(node -> this.nodes.contains(node) && node.isTerminal());
    }

    // TODO(cache?)
    double getAvgOrder() {
        if (nodes.isEmpty())
            return 0;
        double sum = 0;
        for (Vertex node : nodes) {
            sum += node.getOrder();
        }
        return sum / nodes.size();
    }

    @Override
    public String stringRepresentation() {
        StringBuilder sb = new StringBuilder();
        sb.append("Network[order=")
                .append(getAvgOrder())
                .append("] {\n");
        nodes.forEach(node -> sb.append(node.stringRepresentation()).append(",\n"));
        sb.append("}");
        return sb.toString();
    }

    public MutableGraph toGraph() {
        PropertiesHandler cfg = PropertiesHandler.getProps();
        HashMap<Vertex, MutableNode> map = new HashMap<>();
        MutableGraph g = mutGraph("network")
                .setDirected(true);
        nodes.forEach(node -> {
            Color color = cfg.getColor("vertex");
            if (node.isTerminal())
                color = cfg.getColor("terminal");
            if (node.isSelected())
                color = cfg.getColor("selectedN");
            map.put(node, mutNode(node.toString()).add(color).add(Style.FILLED));
        });
        nodes.forEach(node -> {
            MutableNode mNode = map.get(node);
            g.add(mNode);
            node.getConnections().forEach(ch -> {
                MutableNode toNode = map.get(ch.getToNode());
                Color color = cfg.getColor("edge");
                if (ch.isUsed())
                    color = cfg.getColor("connected");
                if (ch.isSelected())
                    color = cfg.getColor("selectedC");
                mNode.addLink(to(toNode).
                        with(Label.markdown(String.valueOf(ch.getWeight())),
                                Style.DASHED,
                                color));
            });
        });
        return g;
    }

    @Override
    public String toString() {
        return "Network[nodes=" + nodes.size() + ", order=" + getAvgOrder() + "]";
    }

    public void closeAll() {
        getChannels(true).forEach(ch -> ch.setUsed(false));
    }

    public void clear() {
        nodes.clear();
        openConnections.clear();
    }
}

File: src/main/java/mvc/controller/Network.kt
package mvc.controller

import guru.nidi.graphviz.attribute.Label
import guru.nidi.graphviz.model.Factory
import guru.nidi.graphviz.model.MutableGraph
import guru.nidi.graphviz.model.MutableNode
import mvc.model.algorithms.BFAFactory
import mvc.model.algorithms.GraphPath
import mvc.model.algorithms.AlgorithmFactory
import mvc.model.channel.Channel
import mvc.model.node.Node
import common.ToInfo
import common.Config

import java.util.*

import guru.nidi.graphviz.model.Factory.*
import mvc.model.channel.SimplexChannelFactory
import mvc.model.node.SimpleNode

class Network @JvmOverloads constructor(private val topology: Topology = object : Topology {
    override fun createNetwork(): MutableList<Node> {
        return ArrayList()
    }

    override fun createNode(): Node {
        return SimpleNode("Node", SimplexChannelFactory())
    }
}, var factory: AlgorithmFactory = BFAFactory()) : ToInfo {
    private var nodes: MutableList<Node> = ArrayList()
    private var openConnections: MutableList<GraphPath> = ArrayList()

    val randomNode: Node
        get() = nodes[Random().nextInt(nodes.size)]

    private val channels: List<Channel>
        get() = nodes
                .flatMap { node -> node.connections }
                .distinct()

    // TODO(cache?)
    internal val avgOrder: Double
        get() {
            if (nodes.isEmpty())
                return .0
            var sum = .0
            for (node in nodes) {
                sum += node.order.toDouble()
            }
            return sum / nodes.size
        }

    init {
        generateNetwork()
    }

    fun generateNetwork() {
        nodes = topology.createNetwork()
        openConnections = ArrayList()
    }

    fun getNodes(terminal: Boolean = false): List<Node> {
        return if (!terminal)
            nodes
        else
            nodes.filter(Node::isTerminal)
    }

    fun getRandomNode(terminal: Boolean): Node {
        val nodes = getNodes(terminal)
        return nodes[Random().nextInt(nodes.size)]
    }

    @JvmOverloads
    fun getPath(from: Node, to: Node, isUsed: Boolean = false): GraphPath {
        return if (!exists(from, to)) GraphPath() else factory.getAlgorithm(nodes, getChannels(isUsed)).getPath(from, to)
    }

    fun getPaths(from: Node): List<GraphPath> {
        val algorithm = factory.getAlgorithm(nodes, channels)
        return nodes
                .filter {  it != from }
                .map {algorithm.getPath(from, it) }
    }

    private fun getChannels(isUsed: Boolean): List<Channel> {
        return nodes
                .flatMap(Node::connections)
                .distinct()
                .filter { it.isUsed == isUsed }
    }

    fun createConnection(from: Node, to: Node): GraphPath {
        if (from === to)
            return GraphPath()
        val path = getPath(from, to)
        if (path.exists()) {
            path.forEach { it.isUsed = true }
            openConnections.add(path)
        }
        return path
    }

    fun closeConnection(path: GraphPath) {
        if (openConnections.contains(path)) {
            path.forEach { ch -> ch.isUsed = false }
            openConnections.remove(path)
        }
    }


    fun closeConnection(from: Node, to: Node) {
        openConnections.stream()
                .filter { path -> path.ofNodes(from, to) }
                .findAny()
                .ifPresent { this.closeConnection(it) }
    }

    fun addNode(): Node {
        val node = topology.createNode()
        nodes.add(node)
        return node
    }

    fun addNode(node: Node) {
        if (!exists(node)) {
            nodes.add(node)
        }
    }

    fun removeNode(node: Node) {
        val connections = node.connections.toTypedArray()
        for (ch in connections) {
            ch.remove()
        }
        nodes.remove(node)
    }

    fun addConnection(from: Node, to: Node, weight: Int) {
        if (exists(from, to)) {
            from.addConnectedNode(to, weight)
        }
    }

    fun removeConnection(from: Node, to: Node) {
        if (exists(from, to)) {
            from.removeConnectedNode(to)
        }
    }

    private fun exists(vararg nodes: Node): Boolean {
        return nodes.all { it in nodes }
    }

    override fun toInfo(): String {
        val sb = StringBuilder()
        sb.append("Network[order=")
                .append(avgOrder)
                .append("] {\n")
        nodes.forEach { sb.append(it.toInfo()).append(",\n") }
        sb.append("}")
        return sb.toString()
    }

    fun toGraph(): MutableGraph {
        val cfg = Config.config
        val map = HashMap<Node, MutableNode>()
        val g = mutGraph("mvc/controller")
                .setDirected(true)
        nodes.forEach { node ->
            var color = cfg.getColor("node")
            if (node.isTerminal)
                color = cfg.getColor("terminal")
            if (node.isSelected)
                color = cfg.getColor("selectedN")
            map[node] = mutNode(node.toString()).add(color)
        }
        nodes.forEach { node ->
            val mNode = map[node]!!
            g.add(mNode)
            node.connections.forEach { ch ->
                val toNode = map[ch.toNode]!!
                var color = cfg.getColor("channel")
                if (ch.isUsed)
                    color = cfg.getColor("connected")
                if (ch.isSelected)
                    color = cfg.getColor("selectedC")
                mNode.addLink(Factory.to(toNode).with(Label.markdown(ch.weight.toString()),
                        color))
            }
        }
        return g
    }

    override fun toString(): String {
        return "Network[nodes=" + nodes.size + ", order=" + avgOrder + "]"
    }

    fun closeAll() {
        getChannels(true).forEach { ch -> ch.isUsed = false }
    }

    fun clear() {
        nodes.clear()
        openConnections.clear()
    }
}
